### DISCUSSION TITLE
how to require luarocks installed modules?

### DISCUSSION THREAD

**Discussion by lplume** [lyte2d/lyte2d#47] [Talk]
I'm having trouble to require modules installed via luarocks; using the lates release binary, linux

```
luarocks install luafilesystem

# in app.lua
local lfs = require("lfs")

lyte .
```

```Warning: could not load file 'config.lua', error: 'not found'
Warning: could not load file 'config/init.lua', error: 'not found'
Warning: could not load file 'config.fnl', error: 'not found'
Warning: could not load file 'config/init.fnl', error: 'not found'
Warning: could not load file 'config.tl', error: 'not found'
Warning: could not load file 'config/init.tl', error: 'not found'
Failed to mount zip './lyte' APP_EXE / (fetched async): PHYSFS: 0 6, unsupported
Warning: could not load file 'lfs.lua', error: 'not found'
Warning: could not load file 'lfs/init.lua', error: 'not found'
Warning: could not load file 'lfs.fnl', error: 'not found'
Warning: could not load file 'lfs/init.fnl', error: 'not found'
Warning: could not load file 'lfs.tl', error: 'not found'
Warning: could not load file 'lfs/init.tl', error: 'not found'
Error: lyte_boot.lua:199: attempt to concatenate local 'lang' (a nil value)
```

is there a way or some env to be put up in place?

thanks

**Comment by morew4rd** [comment]
Unfortunately, that scenario is not supported. The idea is that all the games are distributed as self-contained objects with very little (if any) dependencies on the users' systems. So I have overridden standard module resolution functionality. 

Here are the current Lyte2D restrictions regarding the filesystem access:

- You can't load Lua modules outside of top-level folder. (Well, you can "workaround" by loading it as a text file and "eval"ing. But it's a harder to use foot-gun :))
- You can't load binary files (audio/image etc) outside of top-level folder with lyte APIs.
- You can't write files outside of a pre-described set of locations. (toplevel only currently. users's temporary files etc. are an option for the future.)

For pure Lua modules, this is not a big problem. You can just put the module Lua files/folder at your top-level somewhere and require it from there instead.

This set of restrictions enables developers to distribute their game as a self-contained unit that would work on almost everywhere.

LuaFileSystem library is not pure Lua and also has native code in it, so there's no straightforward way to use it right now. It's doable but would require some gluing (and possibly sanding. :))

Is there some specific functionality you wanted to get out of this library, or was this a general question? If former, Lyte2D filesystem APIs will grow a bit for sure. I'm using PHYSFS library, see some of the APIs that I maybe able to bring in here: https://icculus.org/physfs/docs/html/physfs_8h.html)

At some point, adding native code will be easier but no designs for that yet.

**Comment by lplume** [comment]
Hello! I see the point of the self-contained and 100% second the idea (well I guess it will be that way anyway ^^')!

No specific functionality about lfs that would benefit lyte2d itself, more of a general question plus I was working on #4 (both the imgui and the editor) and found myself i little bit stuck on how to operate on directories and files, but i think i might work out something.

Btw would appreciate if you want to revive the issue adding ideas, i might have a bare minimum text editor in lyte2d published in some way soon(ish)™.

This issue do not need more attention, i'll close it.

thanks

**Comment by morew4rd** [comment]
Very cool! Looking forward to seeing your work! I added some thoughts on that issue and revived it.

Please file more issues, as I'm sure you're hitting more problems while developing the library/editor :) I'd love to hear your thoughts on even the basic mundane stuff.

Thanks again for filing this issue and the chat :)

**Comment by morew4rd** [comment]
Converted this to a discussion (instead of closing), in case other folks wonder about similar things and do a search.
